crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -4684,6 +4708,10 @@ impl<'db> TypeInferenceBuilder<'db> {
             symbol_from_bindings(db, use_def.bindings_at_use(use_id))
         };
 
+        if let Some(use_id) = name_node.try_scoped_use_id(db, scope) {
+            use_ids.push((file_scope_id, use_id));
+        }
```

Comment:
I _think_ putting this here means we are doing unnecessary work? If we don't end up walking up enclosing scopes, we don't need this; the visible binding(s) in this scope will already have narrowing constraints attached to them. So this could be moved down below line 4738.

Reply:
This part is still necessary, because we need to be able to apply narrowing constraints introduced in the current scope to symbols in the outer scope as well.

```python
def f(x: str | None):
    def _():
        # x is not defined in this scope
        if x is not None:
            reveal_type(x)  # revealed: str
```

Reply:
Yes, I know. What I'm saying is that we only need this if the symbol is (maybe) not bound in the current scope and we will find it in an enclosing scope. That is, if we enter the `or_fall_back_to` below, and make it past the check on line 4726. So it seems like we don't need to add this here, we can add it below instead, and avoid adding it in cases where we don't need it.

Reply:
Ah, finally I get what you mean!
Moved.